### PR TITLE
Lazily load nbconvert export formats, and optionally cache results

### LINKED ISSUE #13833: JupyterLab requests /api/nbconvert twice during load
## Description

JupyterLab during webpage load performs requests to /api/nbconvert twice.

## Reproduce

1. Open Chrome DevTools, switch to Network tab
2. Reload the page with JupyterLab
3. Find /api/nbconvert requests (you could filter by 'nbconvert')
4. See there are two simultanous requests

## Expected behavior

I probably would expect one request not two. The /api/nbconvert response time can be relatively slow (is quite slow in my full, non 'jlab-test' environment for some reason). 

## Context

- Operating System and version: CentOS 
- Browser and version: Edge 108.0.1462
- JupyterLab version: 3.5.2

<details><summary>Troubleshoot Output</summary>
<pre>
$ jupyter troubleshoot
$PATH:
        /home/kungurov/.conda/envs/jlab-test/bin
        /usr/frog/scicomp/pythonds/conda/condabin
        /usr/frog/scicomp/pythonds/v1.2/bin
        /usr/frog/scicomp/pythonds/latex/bin/x86_64-linux
        /usr/frog/scicomp/pythonds/conda/condabin
        /usr/local/bin
        /usr/bin
        /usr/local/sbin
        /usr/sbin
        /home/kungurov/bin

sys.path:
        /home/kungurov/.conda/envs/jlab-test/bin
        /home/kungurov/.conda/envs/jlab-test/lib/python311.zip
        /home/kungurov/.conda/envs/jlab-test/lib/python3.11
        /home/kungurov/.conda/envs/jlab-test/lib/python3.11/lib-dynload
        /home/kungurov/.conda/envs/jlab-test/lib/python3.11/site-packages

sys.executable:
        /home/kungurov/.conda/envs/jlab-test/bin/python

sys.version:
        3.11.0 | packaged by conda-forge | (main, Jan 14 2023, 12:27:40) [GCC 11.3.0]

platform.platform():
        Linux-3.10.0-1160.76.1.el7.x86_64-x86_64-with-glibc2.17

which -a jupyter:
        /home/kungurov/.conda/envs/jlab-test/bin/jupyter

pip list:
        Package                       Version
        ----------------------------- -----------
        anyio                         3.6.2
        argon2-cffi                   21.3.0
        argon2-cffi-bindings          21.2.0
        asttokens                     2.2.1
        attrs                         22.2.0
        Babel                         2.11.0
        backcall                      0.2.0
        backports.functools-lru-cache 1.6.4
        beautifulsoup4                4.11.1
        bleach                        5.0.1
        brotlipy                      0.7.0
        certifi                       2022.12.7
        cffi                          1.15.1
        charset-normalizer            2.1.1
        comm                          0.1.2
        cryptography                  39.0.0
        debugpy                       1.6.5
        decorator                     5.1.1
        defusedxml                    0.7.1
        entrypoints                   0.4
        executing                     1.2.0
        fastjsonschema                2.16.2
        flit_core                     3.8.0
        idna                          3.4
        importlib-metadata            6.0.0
        importlib-resources           5.10.2
        ipykernel                     6.20.2
        ipython                       8.8.0
        ipython-genutils              0.2.0
        jedi                          0.18.2
        Jinja2                        3.1.2
        json5                         0.9.5
        jsonschema                    4.17.3
        jupyter_client                7.4.9
        jupyter_core                  5.1.3
        jupyter-events                0.6.3
        jupyter_server                2.1.0
        jupyter_server_terminals      0.4.4
        jupyterlab                    3.5.2
        jupyterlab-pygments           0.2.2
        jupyterlab_server             2.19.0
        MarkupSafe                    2.1.2
        matplotlib-inline             0.1.6
        mistune                       2.0.4
        nbclassic                     0.4.8
        nbclient                      0.7.2
        nbconvert                     7.2.8
        nbformat                      5.7.3
        nest-asyncio                  1.5.6
        notebook                      6.5.2
        notebook_shim                 0.2.2
        packaging                     23.0
        pandocfilters                 1.5.0
        parso                         0.8.3
        pexpect                       4.8.0
        pickleshare                   0.7.5
        pip                           22.3.1
        pkgutil_resolve_name          1.3.10
        platformdirs                  2.6.2
        prometheus-client             0.15.0
        prompt-toolkit                3.0.36
        psutil                        5.9.4
        ptyprocess                    0.7.0
        pure-eval                     0.2.2
        pycparser                     2.21
        Pygments                      2.14.0
        pyOpenSSL                     23.0.0
        pyrsistent                    0.19.3
        PySocks                       1.7.1
        python-dateutil               2.8.2
        python-json-logger            2.0.4
        pytz                          2022.7.1
        PyYAML                        6.0
        pyzmq                         25.0.0
        requests                      2.28.2
        rfc3339-validator             0.1.4
        rfc3986-validator             0.1.1
        Send2Trash                    1.8.0
        setuptools                    66.1.1
        six                           1.16.0
        sniffio                       1.3.0
        soupsieve                     2.3.2.post1
        stack-data                    0.6.2
        terminado                     0.17.1
        tinycss2                      1.2.1
        tomli                         2.0.1
        tornado                       6.2
        traitlets                     5.8.1
        typing_extensions             4.4.0
        urllib3                       1.26.14
        wcwidth                       0.2.6
        webencodings                  0.5.1
        websocket-client              1.4.2
        wheel                         0.38.4
        zipp                          3.11.0

conda list:
        # packages in environment at /home/kungurov/.conda/envs/jlab-test:
        #
        # Name                    Version                   Build  Channel
        _libgcc_mutex             0.1                 conda_forge    defaults
        _openmp_mutex             4.5                       2_gnu    defaults
        anyio                     3.6.2              pyhd8ed1ab_0    defaults
        argon2-cffi               21.3.0             pyhd8ed1ab_0    defaults
        argon2-cffi-bindings      21.2.0          py311hd4cff14_3    defaults
        asttokens                 2.2.1              pyhd8ed1ab_0    defaults
        attrs                     22.2.0             pyh71513ae_0    defaults
        babel                     2.11.0             pyhd8ed1ab_0    defaults
        backcall                  0.2.0              pyh9f0ad1d_0    defaults
        backports                 1.0                pyhd8ed1ab_3    defaults
        backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    defaults
        beautifulsoup4            4.11.1             pyha770c72_0    defaults
        bleach                    5.0.1              pyhd8ed1ab_0    defaults
        brotlipy                  0.7.0           py311hd4cff14_1005    defaults
        bzip2                     1.0.8                h7f98852_4    defaults
        ca-certificates           2022.12.7            ha878542_0    defaults
        certifi                   2022.12.7          pyhd8ed1ab_0    defaults
        cffi                      1.15.1          py311h409f033_3    defaults
        charset-normalizer        2.1.1              pyhd8ed1ab_0    defaults
        comm                      0.1.2              pyhd8ed1ab_0    defaults
        cryptography              39.0.0          py311h9b4c7bb_0    defaults
        debugpy                   1.6.5           py311hcafe171_0    defaults
        decorator                 5.1.1              pyhd8ed1ab_0    defaults
        defusedxml                0.7.1              pyhd8ed1ab_0    defaults
        entrypoints               0.4                pyhd8ed1ab_0    defaults
        executing                 1.2.0              pyhd8ed1ab_0    defaults
        flit-core                 3.8.0              pyhd8ed1ab_0    defaults
        idna                      3.4                pyhd8ed1ab_0    defaults
        importlib-metadata        6.0.0              pyha770c72_0    defaults
        importlib_resources       5.10.2             pyhd8ed1ab_0    defaults
        ipykernel                 6.20.2             pyh210e3f2_0    defaults
        ipython                   8.8.0              pyh41d4057_0    defaults
        ipython_genutils          0.2.0                      py_1    defaults
        jedi                      0.18.2             pyhd8ed1ab_0    defaults
        jinja2                    3.1.2              pyhd8ed1ab_1    defaults
        json5                     0.9.5              pyh9f0ad1d_0    defaults
        jsonschema                4.17.3             pyhd8ed1ab_0    defaults
        jupyter_client            7.4.9              pyhd8ed1ab_0    defaults
        jupyter_core              5.1.3           py311h38be061_0    defaults
        jupyter_events            0.6.3              pyhd8ed1ab_0    defaults
        jupyter_server            2.1.0              pyhd8ed1ab_0    defaults
        jupyter_server_terminals  0.4.4              pyhd8ed1ab_1    defaults
        jupyterlab                3.5.2              pyhd8ed1ab_0    defaults
        jupyterlab_pygments       0.2.2              pyhd8ed1ab_0    defaults
        jupyterlab_server         2.19.0             pyhd8ed1ab_0    defaults
        ld_impl_linux-64          2.39                 hcc3a1bd_1    defaults
        libffi                    3.4.2                h7f98852_5    defaults
        libgcc-ng                 12.2.0              h65d4601_19    defaults
        libgomp                   12.2.0              h65d4601_19    defaults
        libnsl                    2.0.0                h7f98852_0    defaults
        libsodium                 1.0.18               h36c2ea0_1    defaults
        libsqlite                 3.40.0               h753d276_0    defaults
        libstdcxx-ng              12.2.0              h46fd767_19    defaults
        libuuid                   2.32.1            h7f98852_1000    defaults
        libzlib                   1.2.13               h166bdaf_4    defaults
        markupsafe                2.1.2           py311h2582759_0    defaults
        matplotlib-inline         0.1.6              pyhd8ed1ab_0    defaults
        mistune                   2.0.4              pyhd8ed1ab_0    defaults
        nbclassic                 0.4.8              pyhd8ed1ab_0    defaults
        nbclient                  0.7.2              pyhd8ed1ab_0    defaults
        nbconvert                 7.2.8              pyhd8ed1ab_0    defaults
        nbconvert-core            7.2.8              pyhd8ed1ab_0    defaults
        nbconvert-pandoc          7.2.8              pyhd8ed1ab_0    defaults
        nbformat                  5.7.3              pyhd8ed1ab_0    defaults
        ncurses                   6.3                  h27087fc_1    defaults
        nest-asyncio              1.5.6              pyhd8ed1ab_0    defaults
        notebook                  6.5.2              pyha770c72_1    defaults
        notebook-shim             0.2.2              pyhd8ed1ab_0    defaults
        openssl                   3.0.7                h0b41bf4_1    defaults
        packaging                 23.0               pyhd8ed1ab_0    defaults
        pandoc                    2.19.2               h32600fe_1    defaults
        pandocfilters             1.5.0              pyhd8ed1ab_0    defaults
        parso                     0.8.3              pyhd8ed1ab_0    defaults
        pexpect                   4.8.0              pyh1a96a4e_2    defaults
        pickleshare               0.7.5                   py_1003    defaults
        pip                       22.3.1             pyhd8ed1ab_0    defaults
        pkgutil-resolve-name      1.3.10             pyhd8ed1ab_0    defaults
        platformdirs              2.6.2              pyhd8ed1ab_0    defaults
        prometheus_client         0.15.0             pyhd8ed1ab_0    defaults
        prompt-toolkit            3.0.36             pyha770c72_0    defaults
        psutil                    5.9.4           py311hd4cff14_0    defaults
        ptyprocess                0.7.0              pyhd3deb0d_0    defaults
        pure_eval                 0.2.2              pyhd8ed1ab_0    defaults
        pycparser                 2.21               pyhd8ed1ab_0    defaults
        pygments                  2.14.0             pyhd8ed1ab_0    defaults
        pyopenssl                 23.0.0             pyhd8ed1ab_0    defaults
        pyrsistent                0.19.3          py311h2582759_0    defaults
        pysocks                   1.7.1              pyha2e5f31_6    defaults
        python                    3.11.0          he550d4f_1_cpython    defaults
        python-dateutil           2.8.2              pyhd8ed1ab_0    defaults
        python-fastjsonschema     2.16.2             pyhd8ed1ab_0    defaults
        python-json-logger        2.0.4              pyhd8ed1ab_0    defaults
        python_abi                3.11                    3_cp311    defaults
        pytz                      2022.7.1           pyhd8ed1ab_0    defaults
        pyyaml                    6.0             py311hd4cff14_5    defaults
        pyzmq                     25.0.0          py311hd6ccaeb_0    defaults
        readline                  8.1.2                h0f457ee_0    defaults
        requests                  2.28.2             pyhd8ed1ab_0    defaults
        rfc3339-validator         0.1.4              pyhd8ed1ab_0    defaults
        rfc3986-validator         0.1.1              pyh9f0ad1d_0    defaults
        send2trash                1.8.0              pyhd8ed1ab_0    defaults
        setuptools                66.1.1             pyhd8ed1ab_0    defaults
        six                       1.16.0             pyh6c4a22f_0    defaults
        sniffio                   1.3.0              pyhd8ed1ab_0    defaults
        soupsieve                 2.3.2.post1        pyhd8ed1ab_0    defaults
        stack_data                0.6.2              pyhd8ed1ab_0    defaults
        terminado                 0.17.1             pyh41d4057_0    defaults
        tinycss2                  1.2.1              pyhd8ed1ab_0    defaults
        tk                        8.6.12               h27826a3_0    defaults
        tomli                     2.0.1              pyhd8ed1ab_0    defaults
        tornado                   6.2             py311hd4cff14_1    defaults
        traitlets                 5.8.1              pyhd8ed1ab_0    defaults
        typing-extensions         4.4.0                hd8ed1ab_0    defaults
        typing_extensions         4.4.0              pyha770c72_0    defaults
        tzdata                    2022g                h191b570_0    defaults
        urllib3                   1.26.14            pyhd8ed1ab_0    defaults
        wcwidth                   0.2.6              pyhd8ed1ab_0    defaults
        webencodings              0.5.1                      py_1    defaults
        websocket-client          1.4.2              pyhd8ed1ab_0    defaults
        wheel                     0.38.4             pyhd8ed1ab_0    defaults
        xz                        5.2.6                h166bdaf_0    defaults
        yaml                      0.2.5                h7f98852_2    defaults
        zeromq                    4.3.4                h9c3ff4c_1    defaults
        zipp                      3.11.0             pyhd8ed1ab_0    defaults

conda env:
        name: jlab-test
        channels:
          - defaults
        dependencies:
          - _libgcc_mutex=0.1=conda_forge
          - _openmp_mutex=4.5=2_gnu
          - anyio=3.6.2=pyhd8ed1ab_0
          - argon2-cffi=21.3.0=pyhd8ed1ab_0
          - argon2-cffi-bindings=21.2.0=py311hd4cff14_3
          - asttokens=2.2.1=pyhd8ed1ab_0
          - attrs=22.2.0=pyh71513ae_0
          - babel=2.11.0=pyhd8ed1ab_0
          - backcall=0.2.0=pyh9f0ad1d_0
          - backports=1.0=pyhd8ed1ab_3
          - backports.functools_lru_cache=1.6.4=pyhd8ed1ab_0
          - beautifulsoup4=4.11.1=pyha770c72_0
          - bleach=5.0.1=pyhd8ed1ab_0
          - brotlipy=0.7.0=py311hd4cff14_1005
          - bzip2=1.0.8=h7f98852_4
          - ca-certificates=2022.12.7=ha878542_0
          - certifi=2022.12.7=pyhd8ed1ab_0
          - cffi=1.15.1=py311h409f033_3
          - charset-normalizer=2.1.1=pyhd8ed1ab_0
          - comm=0.1.2=pyhd8ed1ab_0
          - cryptography=39.0.0=py311h9b4c7bb_0
          - debugpy=1.6.5=py311hcafe171_0
          - decorator=5.1.1=pyhd8ed1ab_0
          - defusedxml=0.7.1=pyhd8ed1ab_0
          - entrypoints=0.4=pyhd8ed1ab_0
          - executing=1.2.0=pyhd8ed1ab_0
          - flit-core=3.8.0=pyhd8ed1ab_0
          - idna=3.4=pyhd8ed1ab_0
          - importlib-metadata=6.0.0=pyha770c72_0
          - importlib_resources=5.10.2=pyhd8ed1ab_0
          - ipykernel=6.20.2=pyh210e3f2_0
          - ipython=8.8.0=pyh41d4057_0
          - ipython_genutils=0.2.0=py_1
          - jedi=0.18.2=pyhd8ed1ab_0
          - jinja2=3.1.2=pyhd8ed1ab_1
          - json5=0.9.5=pyh9f0ad1d_0
          - jsonschema=4.17.3=pyhd8ed1ab_0
          - jupyter_client=7.4.9=pyhd8ed1ab_0
          - jupyter_core=5.1.3=py311h38be061_0
          - jupyter_events=0.6.3=pyhd8ed1ab_0
          - jupyter_server=2.1.0=pyhd8ed1ab_0
          - jupyter_server_terminals=0.4.4=pyhd8ed1ab_1
          - jupyterlab=3.5.2=pyhd8ed1ab_0
          - jupyterlab_pygments=0.2.2=pyhd8ed1ab_0
          - jupyterlab_server=2.19.0=pyhd8ed1ab_0
          - ld_impl_linux-64=2.39=hcc3a1bd_1
          - libffi=3.4.2=h7f98852_5
          - libgcc-ng=12.2.0=h65d4601_19
          - libgomp=12.2.0=h65d4601_19
          - libnsl=2.0.0=h7f98852_0
          - libsodium=1.0.18=h36c2ea0_1
          - libsqlite=3.40.0=h753d276_0
          - libstdcxx-ng=12.2.0=h46fd767_19
          - libuuid=2.32.1=h7f98852_1000
          - libzlib=1.2.13=h166bdaf_4
          - markupsafe=2.1.2=py311h2582759_0
          - matplotlib-inline=0.1.6=pyhd8ed1ab_0
          - mistune=2.0.4=pyhd8ed1ab_0
          - nbclassic=0.4.8=pyhd8ed1ab_0
          - nbclient=0.7.2=pyhd8ed1ab_0
          - nbconvert=7.2.8=pyhd8ed1ab_0
          - nbconvert-core=7.2.8=pyhd8ed1ab_0
          - nbconvert-pandoc=7.2.8=pyhd8ed1ab_0
          - nbformat=5.7.3=pyhd8ed1ab_0
          - ncurses=6.3=h27087fc_1
          - nest-asyncio=1.5.6=pyhd8ed1ab_0
          - notebook=6.5.2=pyha770c72_1
          - notebook-shim=0.2.2=pyhd8ed1ab_0
          - openssl=3.0.7=h0b41bf4_1
          - packaging=23.0=pyhd8ed1ab_0
          - pandoc=2.19.2=h32600fe_1
          - pandocfilters=1.5.0=pyhd8ed1ab_0
          - parso=0.8.3=pyhd8ed1ab_0
          - pexpect=4.8.0=pyh1a96a4e_2
          - pickleshare=0.7.5=py_1003
          - pip=22.3.1=pyhd8ed1ab_0
          - pkgutil-resolve-name=1.3.10=pyhd8ed1ab_0
          - platformdirs=2.6.2=pyhd8ed1ab_0
          - prometheus_client=0.15.0=pyhd8ed1ab_0
          - prompt-toolkit=3.0.36=pyha770c72_0
          - psutil=5.9.4=py311hd4cff14_0
          - ptyprocess=0.7.0=pyhd3deb0d_0
          - pure_eval=0.2.2=pyhd8ed1ab_0
          - pycparser=2.21=pyhd8ed1ab_0
          - pygments=2.14.0=pyhd8ed1ab_0
          - pyopenssl=23.0.0=pyhd8ed1ab_0
          - pyrsistent=0.19.3=py311h2582759_0
          - pysocks=1.7.1=pyha2e5f31_6
          - python=3.11.0=he550d4f_1_cpython
          - python-dateutil=2.8.2=pyhd8ed1ab_0
          - python-fastjsonschema=2.16.2=pyhd8ed1ab_0
          - python-json-logger=2.0.4=pyhd8ed1ab_0
          - python_abi=3.11=3_cp311
          - pytz=2022.7.1=pyhd8ed1ab_0
          - pyyaml=6.0=py311hd4cff14_5
          - pyzmq=25.0.0=py311hd6ccaeb_0
          - readline=8.1.2=h0f457ee_0
          - requests=2.28.2=pyhd8ed1ab_0
          - rfc3339-validator=0.1.4=pyhd8ed1ab_0
          - rfc3986-validator=0.1.1=pyh9f0ad1d_0
          - send2trash=1.8.0=pyhd8ed1ab_0
          - setuptools=66.1.1=pyhd8ed1ab_0
          - six=1.16.0=pyh6c4a22f_0
          - sniffio=1.3.0=pyhd8ed1ab_0
          - soupsieve=2.3.2.post1=pyhd8ed1ab_0
          - stack_data=0.6.2=pyhd8ed1ab_0
          - terminado=0.17.1=pyh41d4057_0
          - tinycss2=1.2.1=pyhd8ed1ab_0
          - tk=8.6.12=h27826a3_0
          - tomli=2.0.1=pyhd8ed1ab_0
          - tornado=6.2=py311hd4cff14_1
          - traitlets=5.8.1=pyhd8ed1ab_0
          - typing-extensions=4.4.0=hd8ed1ab_0
          - typing_extensions=4.4.0=pyha770c72_0
          - tzdata=2022g=h191b570_0
          - urllib3=1.26.14=pyhd8ed1ab_0
          - wcwidth=0.2.6=pyhd8ed1ab_0
          - webencodings=0.5.1=py_1
          - websocket-client=1.4.2=pyhd8ed1ab_0
          - wheel=0.38.4=pyhd8ed1ab_0
          - xz=5.2.6=h166bdaf_0
          - yaml=0.2.5=h7f98852_2
          - zeromq=4.3.4=h9c3ff4c_1
          - zipp=3.11.0=pyhd8ed1ab_0
        prefix: /home/kungurov/.conda/envs/jlab-test
</pre>
</details>

I think it's because getExportFormats() called twice in the code (see below).

index.js, line 340:
```
void services.nbconvert.getExportFormats().then(response => {
    if (response) {
        const formatLabels = Private.getFormatLabels(translato);
        // Convert export list to palette and menu items. 
```
index.js, line 510:
```
void services.nbconvert.getExportFormats().then(response => {
    if (response) {
        /**
... ... ...
         */
        const rawFormatExclude = [
```

### PR DESCRIPTION
**Pull request by bollwyvl**
## References

- fixes #13833

## Code changes

- `services`
  - only ever request the nbconvert export format list once unless `force`d
- `notebook-extension`
  - defer requesting formats until a notebook is opened
    - do _not_ `force` retry, to maintain existing behavior

## User-facing changes

- the main moon animation will clear faster
  - while the number of requests is still quite large (170) in `dev-mode`
    - the double `nbconvert` request launches new threads, inspects (too many) `entry_points` and generally slows other requests down
- the first time notebook is launched
  - a single call to `/api/nbconvert` will be made
  - there might be a _slight_ delay in populating the notebook tools and commands
    - however, a fresh load of a notebook only incurs about 10 uncacheable requests, and seems to not be meaningfully delayed
    - these UI elements are usually hidden in the UI anyway
- if a notebook is already queued for being opened by a workspace
  - the nbconvert request is _still_ delayed until the widget actually loads (faster)

This preserves a  _potential_ latent "bug", namely that after `pip install`ing a package that provides a new nbconvert exporter, it will not appear in either UI location. However, after `pip install`ing something like that, it doesn't seem like a browser refresh is _entirely_ inappropriate to expect, as very possibly such a thing might also change the UI (e.g. to update a metadata form) which is _also_ not evented in any meaningful way.

## Backwards-incompatible changes

- n/a

## Future Work
- an upstream PR to `jupyter_server` could _potentially_ improve the request time without further slowing down system startup

### CODE REVIEW

**Comment by fcollonval** [comment]
bot please update documentation snapshots

File: packages/services/src/nbconvert/index.ts [hunk 1]
```diff
@@ -28,9 +30,11 @@ export class NbConvertManager {
   readonly serverSettings: ServerConnection.ISettings;
 
   /**
-   * Get whether the application should be built.
+   * Fetch and cache the export formats from the expensive nbconvert handler.
    */
-  async getExportFormats(): Promise<NbConvertManager.IExportFormats> {
+  async fetchExportFormats(): Promise<NbConvertManager.IExportFormats> {
```
Comment (fcollonval) [inline packages/services/src/nbconvert/index.ts:35]: Do we need this to be public?

Comment (bollwyvl) [reply packages/services/src/nbconvert/index.ts:35]: Not strictly.

Comment (bollwyvl) [reply packages/services/src/nbconvert/index.ts]: Made `protected`, so that it can still be overridden without hacks.

**Review by fcollonval — APPROVED** [review]
Thanks @bollwyvl

**Comment by fcollonval** [comment]
bot please update documentation snapshots

**Comment by jtpio** [comment]
CI all green.

Thanks!

**Pull request merged**